In [1]:
# Install dependencies as needed:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load
file_path = "data.csv"

# Load the latest version
df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "uciml/breast-cancer-wisconsin-data",
  file_path,
  # Provide any additional arguments like
  # sql_query or pandas_kwargs. See the
  # documenation for more information:
  # https://github.com/Kaggle/kagglehub/blob/main/README.md#kaggledatasetadapterpandas
)



ModuleNotFoundError: No module named 'kagglehub'

## Exploration des données

Examiner les premières lignes du jeu de données, vérifier les informations générales (types de données, valeurs manquantes), et visualiser la distribution de la variable cible.


In [ ]:
display(df.head())

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

**Reasoning**:
Print the number of unique values for each column to identify columns with constant or near-constant values and calculate the count of each unique value in the 'diagnosis' column to understand the distribution of the target variable.



In [ ]:
print(df.nunique())
print(df['diagnosis'].value_counts())

id                         569
diagnosis                    2
radius_mean                456
texture_mean               479
perimeter_mean             522
area_mean                  539
smoothness_mean            474
compactness_mean           537
concavity_mean             537
concave points_mean        542
symmetry_mean              432
fractal_dimension_mean     499
radius_se                  540
texture_se                 519
perimeter_se               533
area_se                    528
smoothness_se              547
compactness_se             541
concavity_se               533
concave points_se          507
symmetry_se                498
fractal_dimension_se       545
radius_worst               457
texture_worst              511
perimeter_worst            514
area_worst                 544
smoothness_worst           411
compactness_worst          529
concavity_worst            539
concave points_worst       492
symmetry_worst             500
fractal_dimension_worst    535
Unnamed:

## Préparation des données

Gérer les valeurs manquantes si nécessaire, encoder la variable cible en valeurs numériques (0 et 1), et diviser les données en ensembles d'entraînement et de test.



Supprimez la colonne 'Unnamed: 32', encodez la colonne 'diagnosis', séparez les caractéristiques et la cible, et divisez les données en ensembles d'entraînement et de test.





In [ ]:
# Drop the 'Unnamed: 32' column
df = df.drop('Unnamed: 32', axis=1)

# Encode the 'diagnosis' column (M=1, B=0)
df['diagnosis'] = df['diagnosis'].map({'M': 1, 'B': 0})

# Separate features (X) and target (y)
X = df.drop(['id', 'diagnosis'], axis=1)
y = df['diagnosis']

# Split data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (455, 30)
Shape of X_test: (114, 30)
Shape of y_train: (455,)
Shape of y_test: (114,)


## Application de la régression logistique

Entraîner un modèle de régression logistique sur l'ensemble d'entraînement.


In [ ]:
from sklearn.linear_model import LogisticRegression

# Instantiate the model
model = LogisticRegression(max_iter=200) # Increased max_iter for convergence

# Train the model
model.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=200)

## Évaluation du modèle

Évaluer les performances du modèle sur l'ensemble de test à l'aide de métriques appropriées (par exemple, précision, rappel, score F1).


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.9561
Precision: 0.9750
Recall: 0.9070
F1-score: 0.9398


## Interprétation du modèle

Analyser les coefficients du modèle pour comprendre l'influence des différentes caractéristiques sur la prédiction.


In [ ]:
import pandas as pd

# Get the coefficients
coefficients = model.coef_[0]

# Create a Pandas Series to associate coefficients with feature names
feature_importance = pd.Series(coefficients, index=X_train.columns)

# Sort the coefficients by their absolute values in descending order
sorted_feature_importance = feature_importance.abs().sort_values(ascending=False)

# Print the sorted coefficients
print("Sorted Feature Importance (Absolute Value):")
print(feature_importance[sorted_feature_importance.index])

Sorted Feature Importance (Absolute Value):
radius_worst              -1.917812
radius_mean               -1.860202
concavity_worst            1.295138
compactness_worst          1.046318
texture_se                -0.948465
concavity_mean             0.485264
concave points_worst       0.385582
symmetry_worst             0.366751
texture_worst              0.356813
compactness_mean           0.350268
concave points_mean        0.210243
perimeter_mean             0.159457
smoothness_worst           0.133775
texture_mean              -0.113856
symmetry_mean              0.113296
perimeter_worst            0.103732
fractal_dimension_worst    0.098401
concavity_se               0.095603
area_se                    0.092759
perimeter_se              -0.091298
smoothness_mean            0.075719
compactness_se             0.067916
radius_se                 -0.058477
area_worst                 0.031334
concave points_se          0.026785
symmetry_se                0.026740
fractal_dimension_me

## Résumé :

1.  **Expliquez pourquoi la régression logistique peut être appliquée à l'ensemble de données "Breast Cancer Wisconsin (Diagnostic)".**
   La régression logistique peut être appliquée à cet ensemble de données car la tâche est un problème de classification binaire. L'objectif est de prédire si une tumeur est bénigne ou maligne, ce qui correspond à deux résultats distincts. La régression logistique est un algorithme approprié pour modéliser la probabilité d'un résultat binaire basé sur plusieurs variables prédictives.

### Principales conclusions de l'analyse des données

*   L'ensemble de données contient 569 entrées avec 33 colonnes, dont un 'id', un 'diagnosis', de nombreuses caractéristiques numériques et une colonne vide 'Unnamed: 32'.
*   La colonne 'Unnamed: 32' était entièrement vide et a été supprimée.
*   La variable cible 'diagnosis' est binaire, avec 357 cas bénins ('B') et 212 cas malins ('M'), ce qui indique un déséquilibre des classes.
*   La variable cible a été encodée numériquement avec succès, 'M' étant mappé à 1 et 'B' à 0.
*   Les données ont été divisées en ensembles d'entraînement (80 %) et de test (20 %), ce qui a donné 455 échantillons d'entraînement et 114 échantillons de test.
*   Un modèle de régression logistique a été entraîné sur les données d'entraînement.
*   Le modèle entraîné a atteint une précision d'environ 0,9561 sur l'ensemble de test.
*   Le modèle a montré une précision élevée (0,9750) et un bon recall (0,9070).
*   Le score F1 du modèle était de 0,9398.
*   Les caractéristiques avec les coefficients absolus les plus importants, indiquant l'influence la plus significative sur la prédiction, incluent radius_worst, radius_mean, concavity_worst et compactness_worst.

### Perspectives ou prochaines étapes

*   Bien que les performances du modèle soient bonnes, envisagez de traiter le déséquilibre des classes dans les données d'entraînement (par exemple, en utilisant des techniques comme SMOTE) pour potentiellement améliorer le rappel pour la classe minoritaire (Maligne).

*   Examinez plus en détail les caractéristiques ayant les coefficients les plus importants afin de comprendre leur signification biologique et de confirmer leur importance dans le diagnostic du cancer du sein. La mise à l'échelle des caractéristiques pourrait également aider le modèle de régression logistique à converger plus rapidement et pourrait avoir un impact sur l'interprétation des coefficients.

